communication/src/main/java/datadog/communication/fleet/FleetServiceImpl.java

```diff
@@ -0,0 +1,231 @@
+import okhttp3.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FleetServiceImpl implements FleetService {
+
+  private static final Logger log = LoggerFactory.getLogger(FleetServiceImpl.class);
+  private static final String CONFIG_PRODUCT_HEADER = "Datadog-Client-Config-Product";
+  private static final int MAX_RESPONSE_SIZE = 8 * 1024 * 1024; // 8 MB
+
+  private final SharedCommunicationObjects sco;
+  private final Thread thread;
+
+  // for testing
+  volatile CountDownLatch testingLatch;
+
+  private final ConcurrentMap<Product, FleetSubscriptionImpl> subscriptions =
+      new ConcurrentHashMap<>();
+
+  public FleetServiceImpl(SharedCommunicationObjects sco, AgentThreadFactory agentThreadFactory) {
+    this.sco = sco;
+    this.thread = agentThreadFactory.newThread(new AgentConfigPollingRunnable());
+  }
+
+  @Override
+  public void init() {
+    thread.start();
+  }
+
+  @Override
+  public FleetService.FleetSubscription subscribe(
+      Product product, final ConfigurationListener listener) {
+    FleetSubscriptionImpl sub = new FleetSubscriptionImpl(product, listener);
+    subscriptions.put(product, sub);
+    return sub;
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.thread.interrupt();
+    try {
+      this.thread.join(5000);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      log.warn("Interrupted waiting for thread " + this.thread.getName() + "to join");
+    }
+  }
+
+  private class AgentConfigPollingRunnable implements Runnable {
+    private static final double BACKOFF_INITIAL = 3.0d;
+    private static final double BACKOFF_BASE = 3.0d;
+    private static final double BACKOFF_MAX_EXPONENT = 3.0d;
+
+    private int consecutiveFailures;
+    private OkHttpClient okHttpClient;
+    private HttpUrl httpUrl;
+
+    @Override
+    public void run() {
+      this.okHttpClient = sco.okHttpClient;
+      this.httpUrl =
+          sco.agentUrl.newBuilder().addPathSegment("v0.6").addPathSegment("config").build();
+
+      if (testingLatch != null) {
+        testingLatch.countDown();
+      }
+      while (!Thread.interrupted()) {
+        try {
+          boolean success = mainLoopIteration();
+          if (testingLatch != null) {
+            testingLatch.countDown();
+          }
+          if (success) {
+            successWait();
+          } else {
+            failureWait();
+          }
+        } catch (InterruptedException e) {
+          log.info("Interrupted; exiting");
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
+
+    private boolean mainLoopIteration() throws InterruptedException {
+      boolean anySuccess = false;
+      Collection<FleetSubscriptionImpl> subs = subscriptions.values();
+      if (subs.isEmpty()) {
+        return true;
+      }
+
+      for (FleetSubscriptionImpl sub : subs) {
+        anySuccess |= fetchConfig(sub);
+      }
+
+      return anySuccess;
+    }
```

Comment:
I suppose, we can rid off `isEmpty` check
```suggestion
    private boolean mainLoopIteration() throws InterruptedException {
      boolean anySuccess = false;
      for (FleetSubscriptionImpl sub : subscriptions.values()) {
        anySuccess |= fetchConfig(sub);
      }
      return anySuccess;
    }
```

Reply:
With these changes, it would return `false` (failure) if there were no subscriptions.

Reply:
Oh, right!